KEYCLOAK-10933 Interfaces for vault SPI

hmlnarik opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

hmlnarik committed:
KEYCLOAK-10933 Interfaces for vault SPI

abstractj commented on server-spi-private/src/main/java/org/keycloak/vault/VaultProviderFactory.java:
```diff
@@ -0,0 +1,26 @@
+import org.keycloak.provider.ProviderFactory;
+
+/**
+ * @author <a href="mailto:mstrukel@redhat.com">Marko Strukelj</a>
```
@hmlnarik maybe by accident you added the wrong author?

hmlnarik replied:
good catch, thanks :)

abstractj commented on server-spi-private/src/main/java/org/keycloak/vault/VaultSpi.java:
```diff
@@ -0,0 +1,48 @@
+import org.keycloak.provider.Spi;
+
+/**
+ * @author <a href="mailto:mstrukel@redhat.com">Marko Strukelj</a>
```
same thing here

slaskawi commented on server-spi-private/src/main/java/org/keycloak/vault/VaultProvider.java:
```diff
@@ -0,0 +1,44 @@
+     * @return The secret or {@code null} if the secret was successfully resolved, or
+     *         an empty {@link java.util.Optional}
+     */
+    Optional<byte[]> obtainSecret(String vaultSecretId);
```
So the other idea is to put `WeakReference<String>` here.

hmlnarik replied:
Vault generally returns a bit sequence, which in Java the closest object is byte array. `String` is interpretation of this byte sequence. Hence String would only be obtained from that byte array in the transcriber object.

slaskawi commented:
@hmlnarik I'm not sure we should create a PR at this point. Perhaps it would be great to have an end-to-end working prototype (we can all implement this in your repo if you wish) and then create a PR. 

I know, my approach has lots of drawbacks.... but it's pretty hard to envision how all the bits work together without having an end-to-end code in from of the eyes...

hmlnarik commented:
> @hmlnarik I'm not sure we should create a PR at this point. Perhaps it would be great to have an end-to-end working prototype (we can all implement this in your repo if you wish) and then create a PR.
> 
> I know, my approach has lots of drawbacks.... but it's pretty hard to envision how all the bits work together without having an end-to-end code in from of the eyes...

Approach with `String`s does not change anything in the low-level vault API. String interpretation is task for the transcriber.

hmlnarik commented:
I realized that we need to return `Autocloseable<Optional<byte[]>>` instead so that the value can be overwritten with garbage right after use.

slaskawi commented:
I think the `Autocloseable<Optional<byte[]>>` looks a bit scary at the API level. How about creating our own class (might be abstract or might be an interface extending `Autocloseable`) that would do all this stuff?

@sguilhen We could also appreciate your feedback at this point. The functionality you're implementing will use this heavily!

sguilhen commented:
> I realized that we need to return `Autocloseable<Optional<byte[]>>` instead so that the value can be overwritten with garbage right after use.

Do we really? If access to the secrets will go through the transcription object then wouldn't it make sense to have the transcription object obtain the secrets from the provider as byte[] and then enclose them as an Autocloseable instance?

sguilhen commented:
> @sguilhen We could also appreciate your feedback at this point. The functionality you're implementing will use this heavily!

I do like the idea of having another type (could be Secret, Password, w/e) that extends `Autocloseable`. So code using the transcription object would be something like

```java
try (Secret<char[]> secret = VaultTranscription.getCharArraySecret("${vault.smtp-pass}")) {
     char[] smtpSecret = secret.get();
     ... // do whatever is needed with secret
} // at this point the `Autocloseable` secret writes random stuff into the char[]
```

Now, thinking about it, I see Hynek's point of having the provider return an `Autocloseable` instance because the transcription object will need to convert the `byte[]` into other formats (such as the `char[]` above) and override the `byte[]` with random stuff. Something like

```java
public static Secret<char[]> getCharArraySecret(final String vaultString) {
    ... // parse the ${vault.id} string
    try (Secret<byte[]> secretBytes = provider.obtainSecret(id)) {
       char[] secretChars = // convert the byte[] to char[]
       return new Secret(secretChars);
    } // at this point the original byte[] as returned from the vault provider is overriden with random values
```

sguilhen commented:
Also, this is missing the VaultSpi entry in org.keycloak.provider.Spi, no?

slaskawi commented:
> Also, this is missing the VaultSpi entry in org.keycloak.provider.Spi, no?

@sguilhen No worries, I'll be adding it: https://github.com/keycloak/keycloak/pull/6223/files#diff-8860817bcbe7df40def1220b15e28baaR1

hmlnarik commented:
@slaskawi @sguilhen Thanks for your comments!

I have updated the PR with a class and a default implementation of a `VaultRawSecret` which is `Autocloseable`. It looks as the best possible way so that the vault can be easily and safely used in try with resources block. I have put JavaDoc there in this regard to leave less space for programmer mistakes.

> Also, this is missing the VaultSpi entry in org.keycloak.provider.Spi, no?

You are indeed right, thanks, fixed in the PR.

slaskawi commented on server-spi-private/src/main/java/org/keycloak/vault/VaultSpi.java:
```diff
@@ -0,0 +1,47 @@
+import org.keycloak.provider.ProviderFactory;
+import org.keycloak.provider.Spi;
+
+/**
```
Probably it would be good to have a javadoc here.

slaskawi commented on services/src/main/java/org/keycloak/vault/DefaultVaultRawSecret.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class DefaultVaultRawSecret implements VaultRawSecret {
+
+    private final Optional<byte[]> rawSecret;
```
In general, I don't like `Optional`s as field. We've had a lengthy discussion about them in Infinispan. A very quick summary is the following:
- `Optional`s don't implement `Serializable`, their intention is to be used as a return type, not as a field.
- `Optional`s used as fields are extremely likely (unless you're using non-standard JVM parameters) to be JIT'ted out. The C2 compiler eliminates even the object creation.
- When you use `Optional` as a field, it's very likely that it will cost you at least an object allocation. The `Optional` object will stay there until it's `reachable` from GC perspective. Of course, if we assume that `DefaultVaultRawSecret` is a short-lived object, that doesn't apply here.

Of course, I'm not insisting to change it here (I've taken part in a flame war around `Optional`s once and I'm not doing this again). I can give you some links if you wish but the `Optional` was often turns into `tabs vs spaces`-type of war.

But as a final thought - you do check if the `Optional` is present and then if the `buffer` is not null, which would get even simpler is the `buffer` was just a nullable `byte[]`.

hmlnarik replied:
These are good points. I've updated the default implementation accordingly.

slaskawi reviewed (dismissed):
LGTM - only two things to consider but they are not blockers. Fix them if you wish or leave them as is.

mposolda reviewed (approved):
I see in the design, it is mentioned that vault SPI is read-only. So I am approving this PR to not block further work as it mets what is in the design document.

But IMO some writing support for the OOTB implementation (and hence maybe in the SPI itself) will be nice. I just sent an email to keycloak-dev to the thread "File-based Vault implementation" to clarify this.

sguilhen commented on services/src/main/java/org/keycloak/vault/DefaultVaultRawSecret.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Default raw secret implementation for {@code byte[]}.
```
Minor review: this should be updated to {@code ByteBuffer}

hmlnarik commented:
Thanks all for your reviews!
